### ISSUE TITLE
the problem about sc.pp.highly_variable_genes

### ISSUE THREAD

**Issue by yuu342** [scverse/scvi-tools#3742]
I run the procedure but failed. sc.pp.highly_variable_genes(
    adata_sub,
    n_top_genes=2000,
    flavor="seurat_v3",
    layer="counts",
    subset=True,
    batch_key="batch"  # Change depending on the batch key for your dataset
)
The kernel automatically restarted.And when I don't set the batch_key,it can run successfully.So i wonder whether deleting the batch_key can influence the result.And if not, could youplease help me solve the problem.there are 110000 cells and 100 batches.I have retained the genes that are expressed in > 10%cells

**Comment by ori-kron-wis** [comment]
Hard to say what went wrong. what exactly are you trying to do?
Is this a scanpy/anndata issue or scvi-tools issue and if so, why?

**Comment by yuu342** [comment]
Thank you for your reply. I plan to use the scvi-tool to reduce the batch effect.
After selecting the highly variable genes and subset the adata,
I hope to run scvi.model.SCVI.setup_anndata(adata_sub, layer="counts", batch_key="batch")
model = scvi.model.SCVI(adata_sub, n_layers=2, n_latent=30, gene_likelihood="nb").But when I added the batch key ,I can't run the  sc.pp.highly_variable_genes.So I am considering whether I need to set the batch key.

**Comment by ori-kron-wis** [comment]
it still not so clear to me.
are you trying to run HVG after registering the model with the same subset adata?
Usually, we run HVG before that (see our tutorials). Either way it will work. 

It might be that you have a memory issue that caused the script to crash. I can't help you with that, but you can verify if you run a smaller adata of cells and batches. Also, make sure your "counts" layer is populated.

Anyway, something like this should work (verify at your end and adjust):

```
from scvi.data import synthetic_iid
import scanpy as sc
from scvi.model import SCVI

adata = synthetic_iid(n_batches=100,n_genes=10000, n_labels=10,batch_size=1000)
adata.layers["counts"] = adata.X.copy()

sc.pp.highly_variable_genes(
    adata,
    n_top_genes=2000,
    flavor="seurat_v3",
    layer="counts",
    subset=True,
    batch_key="batch" # Change depending on the batch key for your dataset
)

SCVI.setup_anndata(
    adata,
    layer="counts",
    batch_key="batch",
    labels_key="labels",
)
model = SCVI(adata, n_layers=2, n_latent=30, gene_likelihood="nb")

model.train(1, check_val_every_n_epoch=1, train_size=0.5)
model.is_trained
```

**Comment by yuu342** [comment]
Thank you for your reply.
My desired code are as you have mentioned.
adata.layers["counts"] = adata.X.copy()

sc.pp.highly_variable_genes(
    adata,
    n_top_genes=2000,
    flavor="seurat_v3",
    layer="counts",
    subset=True,
    batch_key=‘batch' # Change depending on the batch key for your dataset
)

SCVI.setup_anndata(
    adata,
    layer="counts",
    batch_key="batch",
    labels_key="labels",
)
model = SCVI(adata, n_layers=2, n_latent=30, gene_likelihood="nb")
However,I cannot run the highly_variable_genes,so I cannot use the subseted adata to set up the model and train the model.And considering the memory , can I  set the batch_key =none , and then use the subseted adata to set up the model and train? Like the following code
adata.layers["counts"] = adata.X.copy()

sc.pp.highly_variable_genes(
    adata,
    n_top_genes=2000,
    flavor="seurat_v3",
    layer="counts",
    subset=True,
    batch_key=None # Change depending on the batch key for your dataset
)

SCVI.setup_anndata(
    adata,
    layer="counts",
    batch_key="batch",
    labels_key="labels",
)
model = SCVI(adata, n_layers=2, n_latent=30, gene_likelihood="nb")

**Comment by ori-kron-wis** [comment]
I see. Yes, you should be able to still run despite you didn't set batch_key in scanpy HVG.
Whether this will be worse than running with it, its a quality of your data. So either you get a stronger machine, or run like you already did.

**Comment by yuu342** [comment]
Thank you very much for your explanation and help!

**Comment by ori-kron-wis** [comment]
@yuu342 I will just add that it can also be the result of numerical instability in the Seurat v3 HVG implementation used by Scanpy when fitting the LOESS curve.
If was not added before, this might also help:
`sc.pp.filter_genes(adata, min_counts=5)`
but if that still doesn't work and your dataset is large, atlas scale, yes,` batch=None` is fine

**Comment by yuu342** [comment]
Hello,I tried your method and it still didn't work.When I changed the span to 0.8,then it(the batch_key='batch') did work .Thank you for your suggestions!

**Comment by ori-kron-wis** [comment]
great! It probably means that some of your batch group are very small
